src/boss/lambdastore/store.go

```diff
@@ -188,34 +182,42 @@ func (s *LambdaStore) unregisterTriggers(functionName string) {
 // assumes the caller holds the function lock
 func (s *LambdaStore) addToRegistry(name string, body io.Reader) error {
 	tarPath := filepath.Join(s.StorePath, name+".tar.gz")
+	tmpPath := tarPath + ".tmp"
 
-	tarFile, err := os.Create(tarPath)
+	tarFile, err := os.Create(tmpPath)
 	if err != nil {
-		return fmt.Errorf("failed to create lambda tarball: %w", err)
+		return fmt.Errorf("failed to create temp tarball: %w", err)
 	}
-	defer tarFile.Close()
 
-	// TODO: atomically write tarball by writing to a temp file and renaming
 	if _, err := io.Copy(tarFile, body); err != nil {
-		return fmt.Errorf("failed to write tarball: %w", err)
+		tarFile.Close()
+		_ = os.Remove(tmpPath)
+		return fmt.Errorf("failed to write to temp tarball: %w", err)
 	}
 
-	cfg, err := common.ExtractConfigFromTarGz(tarPath)
+	if err := tarFile.Close(); err != nil {
+		_ = os.Remove(tmpPath)
+		return fmt.Errorf("failed to close temp tarball: %w", err)
+	}
+
+	cfg, err := common.ExtractConfigFromTarGz(tmpPath)
 	if err != nil {
-		// Clean up the bad tarball file
-		_ = os.Remove(tarPath)
-		return fmt.Errorf("failed to extract config: %w", err)
+		_ = os.Remove(tmpPath)
+		return fmt.Errorf("failed to extract config from temp tarball: %w", err)
+	}
+
+	// Atomically replace the old file
+	if err := os.Rename(tmpPath, tarPath); err != nil {
+		_ = os.Remove(tmpPath)
+		return fmt.Errorf("failed to rename temp tarball: %w", err)
 	}
 
 	s.mapLock.Lock()
 	defer s.mapLock.Unlock()
 
 	entry, ok := s.Lambdas[name]
 	if !ok {
```

Comment:
Errors are good for regular stuff that can happen.  If something should never happen (like, we have a bug that breaks assumptions), panic is appropriate.  However, is there a way to avoid this check?  For example, this could be a method on LambdaEntry instead of LambdaStore, so that there is no way to call it with a missing entry.  You could also get rid of getFuncLock -- better to have a function that gets an entry, then use the lock within that entry.